README.md
[![Latest release](https://img.shields.io/github/release/jedisct1/piknik.svg)](https://github.com/jedisct1/piknik/releases/latest)
[![Build status](https://travis-ci.org/jedisct1/piknik.svg?branch=master)](https://travis-ci.org/jedisct1/piknik?branch=master)

# Piknik
Copy/paste anything over the network!

[[watch a demo on Asciinema](https://asciinema.org/a/80708)] -
[[download the source code / binaries](https://github.com/jedisct1/piknik/releases/latest)]

![Piknik](https://raw.github.com/jedisct1/piknik/master/piknik.png)

Ever needed a copy/paste clipboard that works over the network?

Piknik seamlessly and securely transfers URLs, code snippets, documents, virtually anything between arbitrary hosts.

No SSH needed, and hosts can sit behind NAT gateways, on different networks.

Fill in the clipboard ("copy") with whatever comes in to the standard input:

```bash
$ pkc
clipboard content
```

Magically retrieve that content from any other host having Piknik installed with the same configuration:

```bash
$ pkp
clipboard content
```

Boom.

Obviously, it can be used to transfer files as well:

```bash
$ pkc < kitten.gif
$ pkp > kittencopy.gif
```

```bash
$ tar cvf - *.txt | pkc
$ pkp | tar xvf -
```

In order to work around firewalls/NAT gatways, the clipboard content transits over TCP via a staging server.

Nothing transits without end-to-end encryption; the server cannot learn much about what the clipboard actually contains.

Data can be shared between different operating systems, including MacOS, Linux and Windows.

## Installation

### Option 1: use precompiled binaries

Precompiled binaries for MacOS, Linux (i386, x86_64, ARM), Win32, Win64, DragonflyBSD, NetBSD and FreeBSD can be downloaded here:
https://github.com/jedisct1/piknik/releases/latest

### Option 2 (on MacOS): use Homebrew

```bash
$ brew install piknik
```

### Option 3: compile the source code

This project is written in Go. So, a Go compiler is required, as well as the following incantation:

```bash
$ export GOPATH=${GOPATH:-~/go} ; mkdir -p $GOPATH
$ go get github.com/jedisct1/piknik
```

The `piknik` executable file should then be available in `$GOPATH/bin`.

## Setup

Piknik requires a bunch of keys. Generate them all with

```bash
$ piknik -genkeys
```

This generates random keys (highly recommended).

A temporary alternative is to derive the keys from a password. The same password will always generate the same set of keys, on all platforms. In order to do so, add the `-password` switch:

```bash
$ piknik -genkeys -password
```

The output of the `-genkeys` command is all you need to build a configuration file.

Only copy the section for servers on the staging server. Only copy the section for clients on the clients.

Is a host gonna act both as a staging server and as a client? Ponder on it before copying the "hybrid" section, but it's there, just in case.

The default location for the configuration file is `~/.piknik.toml`. With the exception of Windows, where dot-files are not so common. On that platform, the file is simply called `piknik.toml`.

Sample configuration file for a staging server:
```toml
Listen = "0.0.0.0:8075"         # Edit appropriately
Psk    = "bf82bab384697243fbf616d3428477a563e33268f0f2307dd14e7245dd8c995d"
SignPk = "0c41ca9b0a1b5fe4daae789534e72329a93a352a6ad73d6f1d368d8eff37271c"
```

Sample configuration file for clients:
```toml
Connect   = "127.0.0.1:8075"    # Edit appropriately
Psk       = "bf82bab384697243fbf616d3428477a563e33268f0f2307dd14e7245dd8c995d"
SignPk    = "0c41ca9b0a1b5fe4daae789534e72329a93a352a6ad73d6f1d368d8eff37271c"
SignSk    = "cecf1d92052f7ba87da36ac3e4a745b64ade8f9e908e52b4f7cd41235dfe7481"
EncryptSk = "2f530eb85e59c1977fce726df9f87345206f2a3d40bf91f9e0e9eeec2c59a3e4"
```

Do not use these, uh? Get your very own keys with the `piknik -genkeys` command.
Edit the `Connect` and `Listen` properties to reflect the staging server IP and port.
And `chmod 600 ~/.piknik.toml` might not be a bad idea.

Don't like the default config file location? Use the `-config` switch.

## Usage (staging server)

Run the following command on the staging server (or use `runit`, `openrc`, `systemd`, whatever to run it as a background service):

```bash
$ piknik -server
```

The staging server has to be publicly accessible. At the very least, it must be reachable by the clients over TCP with the port you specify in the configuration.

Commands without a valid API key (present in the client configuration file) will be rejected by the server.

## Usage (clients)

```bash
$ piknik -copy
```

Copy the standard input to the clipboard.

```bash
$ piknik -paste
```

Retrieve the content of the clipboard and spit it to the standard output.
`-paste` is actually a no-op. This is the default action if `-copy` was not specified.

```bash
$ piknik -move
```

Retrieve the content of the clipboard, spit it to the standard output
and clear the clipboard. Not necessarily in this order.
Only one lucky client will have the privilege to see the content.

That's it.

Feed it anything. Text, binary data, whatever. As long as it fits in memory.

## Suggested shell aliases

Wait. Where are the `pkc` and `pkp` commands mentioned earlier?

Sample shell aliases:

```bash
# pko <content> : copy <content> to the clipboard
pko() {
    echo "$*" | piknik -copy
}

# pkf <file> : copy the content of <file> to the clipboard
pkf() {
    piknik -copy < $1
}

# pkc : read the content to copy to the clipboard from STDIN
alias pkc='piknik -copy'

# pkp : paste the clipboard content
alias pkp='piknik -paste'

# pkm : move the clipboard content
alias pkm='piknik -move'

# pkz : delete the clipboard content
alias pkz='piknik -copy < /dev/null'
```

## Use cases

Use it to:

- Securely send passwords, API keys, URLs from one host to another
- Share a clipboard with your teammates (which can be a lot of fun)
- Copy data from/to isolated VMs, without the VMWare tools or shared volumes (great for unsupported operating systems and malware sandboxes)
- Copy files from/to a Windows machine, without Samba or SSH
- Transfer data between hosts sitting behind firewalls/NAT gateways
- Easily copy configuration files to multiple hosts
- Start a slow download at the office, retrieve it later at home
- Quickly backup a file to the cloud before messing with it
- ...and more!

## Protocol

Common definitions:
```
k: API key
ek: 256-bit symmetric encryption key
ekid: encryption key id encoded as a 64-bit little endian integer
m: plaintext
ct: XChaCha20 ek,n (m)
Hk,s: BLAKE2b(domain="SK", key=k, salt=s, size=32)
Len(x): x encoded as a 64-bit little endian unsigned integer
n: random 192-bit nonce
r: random 256-bit client nonce
r': random 256-bit server nonce
ts: Unix timestamp as a 64-bit little endian integer
Sig: Ed25519
v: 5
```

Copy:
```
-> v || r || h0
h0 := Hk,0(v || r)

<- v || r' || h1
h1 := Hk,1(v || r' || h0)

-> 'S' || h2 || Len(n || ct) || ekid || ts || s || n || ct
s := Sig(n || ct)
h2 := Hk,2(h1 || 'S' || ekid || ts || s)

<- Hk,3(h2)
```

Move/Paste:
```
Move:  opcode := 'M'
Paste: opcode := 'G'

-> v || r || h0
h0 := Hk,0(v || r)

<- v || r' || h1
h1 := Hk,1(v || r' || H0)

-> opcode || h2
h2 := Hk,2(h1 || opcode)

<- Hk,3(h2 || ekid || ts || s) || Len(n || ct) || ekid || ts || s || n || ct
s := Sig(n || ct)
```

## License

[ISC](https://en.wikipedia.org/wiki/ISC_license).

## Credits

Piknik diagram by [EasyPi](https://easypi.herokuapp.com/copy-paste-anything-over-network/).


ChangeLog
* Version 0.9.1
 - Piknik can now be compiled for Plan9 and Solaris.
 - Sending an INFO signal to the server now prints whether the clipboard
is empty or not.
 - Godep was replaced with Glide.
 - Documentation fixes.

* Version 0.9
 - Support for protocol version 3 (Piknik <= 0.2) was removed from the server.
 Please upgrade the clients running these old versions.
 - Compatibility with protocol version 4 (Piknik >= 0.3 and <= 0.7) was
 restored.

* Version 0.8
 - The clipboard now includes a timestamp, so that clients can reject
content that is too old. By default, the maximum age is 7 days. This
can be adjusted with a "TTL" property in the configuration file. The
duration is expressed in seconds.

* Version 0.7
 - The number of simultaneous connections to a Piknik server can now
be controlled with the `-maxclients` command-line flag. When more than
90% of the available slots are filled, new sessions can only be open by
client IPs having recently completed a successful handshake.
 - New command-line flags: `-timeout` and `-datatimeout` in order to
enforce deadlines for handshakes and data transfers.

* Version 0.6
 - Improved error messages

* Version 0.5
 - A new switch, `-password` can now follow `-genkeys` in order to
deterministically derive the keys from a password. Not recommended,
but it can be useful to create an temporary initial configuration on air
gapped devices.
 - An Homebrew (tap) formula was added to easily install it and keep it up
to date on MacOS.

* Version 0.4
 - Servers recomputed the key ID instead of reading it from the
client, requiring the `EncryptSk` property to be present. This is no
longer the case. Thanks to jpmens@ for the bug report.

* Version 0.3
 - Improved documentation
 - The protocol was slightly changed; old clients are still supported
but the server must be updated in order to communicate with clients
running this version.
 - Precompiled binaries are now available for many architectures.

* Version 0.2
 - New `-move` operation to paste the clipboard content and delete it
afterwards. This introduces a protocol change, requiring an update to
the server and to the clients.

* Version 0.1
 - Initial public release.


piknik.go
package main

import (
	"encoding/binary"
	"encoding/hex"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"runtime"
	"time"

	"github.com/BurntSushi/toml"
	blake2b "github.com/minio/blake2b-simd"
	"github.com/mitchellh/go-homedir"
)

const (
	// Version - Piknik version
	Version = "0.9.1"
	// DomainStr - BLAKE2 domain (personalization)
	DomainStr = "PK"
	// DefaultListen - Default value for the Listen parameter
	DefaultListen = "0.0.0.0:8075"
	// DefaultConnect - Default value for the Connect parameter
	DefaultConnect = "127.0.0.1:8075"
	// DefaultTTL - Time after the clipboard is considered obsolete, in seconds
	DefaultTTL = 7 * 24 * time.Hour
)

type tomlConfig struct {
	Connect     string
	Listen      string
	EncryptSk   string
	EncryptSkID uint64
	Psk         string
	SignPk      string
	SignSk      string
	Timeout     uint
	DataTimeout uint
	TTL         uint
}

// Conf - Shared config
type Conf struct {
	Connect        string
	Listen         string
	MaxClients     uint64
	MaxLen         uint64
	EncryptSk      []byte
	EncryptSkID    []byte
	Psk            []byte
	SignPk         []byte
	SignSk         []byte
	Timeout        time.Duration
	DataTimeout    time.Duration
	TTL            time.Duration
	TrustedIPCount uint64
}

func expandConfigFile(path string) string {
	file, err := homedir.Expand(path)
	if err != nil {
		log.Fatal(err)
	}
	return file
}

func version() {
	fmt.Printf("\nPiknik v%v (protocol version: %v)\n",
		Version, DefaultClientVersion)
}

func confCheck(conf Conf, isServer bool) {
	if len(conf.Psk) != 32 {
		log.Fatal("Configuration error: the Psk property is either missing or invalid")
	}
	if len(conf.SignPk) != 32 {
		log.Fatal("Configuration error: the SignPk property is either missing or invalid")
	}
	if isServer {
		if len(conf.Listen) < 3 {
			log.Fatal("Configuration error: the Listen property must be valid for a server")
		}
		if conf.MaxClients <= 0 {
			log.Fatal("Configuration error: MaxClients should be at least 1")
		}
	} else {
		if len(conf.Connect) < 3 {
			log.Fatal("Configuration error: the Connect property must be valid for a client")
		}
		if len(conf.EncryptSk) != 32 || len(conf.SignSk) != 64 {
			log.Fatal("Configuration error: the EncryptSk and SignSk properties must be present\n" +
				"and valid in order to use this command in client mode")
		}
		if conf.TTL <= 0 {
			log.Fatal("TTL cannot be 0")
		}
	}
}

func main() {
	log.SetFlags(0)

	isCopy := flag.Bool("copy", false, "store content (copy)")
	_ = flag.Bool("paste", false, "retrieve the content (paste) - this is the default action")
	isMove := flag.Bool("move", false, "retrieve and delete the clipboard content")
	isServer := flag.Bool("server", false, "start a server")
	isGenKeys := flag.Bool("genkeys", false, "generate keys")
	isDeterministic := flag.Bool("password", false, "derive the keys from a password (default=random keys)")
	maxClients := flag.Uint64("maxclients", 10, "maximum number of simultaneous client connections")
	maxLenMb := flag.Uint64("maxlen", 0, "maximum content length to accept in Mb (0=unlimited)")
	timeout := flag.Uint("timeout", 10, "connection timeout (seconds)")
	dataTimeout := flag.Uint("datatimeout", 3600, "data transmission timeout (seconds)")
	isVersion := flag.Bool("version", false, "display package version")

	defaultConfigFile := "~/.piknik.toml"
	if runtime.GOOS == "windows" {
		defaultConfigFile = "~/piknik.toml"
	}
	configFile := flag.String("config", defaultConfigFile, "configuration file")
	flag.Parse()
	if *isVersion {
		version()
		return
	}
	tomlData, err := ioutil.ReadFile(expandConfigFile(*configFile))
	if err != nil && *isGenKeys == false {
		log.Fatal(err)
	}
	var tomlConf tomlConfig
	if _, err = toml.Decode(string(tomlData), &tomlConf); err != nil {
		log.Fatal(err)
	}
	var conf Conf
	if tomlConf.Listen == "" {
		conf.Listen = DefaultListen
	} else {
		conf.Listen = tomlConf.Listen
	}
	if tomlConf.Connect == "" {
		conf.Connect = DefaultConnect
	} else {
		conf.Connect = tomlConf.Connect
	}
	if *isGenKeys {
		leKey := ""
		if *isDeterministic {
			leKey = getPassword("Password> ")
		}
		genKeys(conf, *configFile, leKey)
		return
	}
	pskHex := tomlConf.Psk
	psk, err := hex.DecodeString(pskHex)
	if err != nil {
		log.Fatal(err)
	}
	conf.Psk = psk
	if encryptSkHex := tomlConf.EncryptSk; encryptSkHex != "" {
		encryptSk, err := hex.DecodeString(encryptSkHex)
		if err != nil {
			log.Fatal(err)
		}
		conf.EncryptSk = encryptSk
	}
	if signPkHex := tomlConf.SignPk; signPkHex != "" {
		signPk, err := hex.DecodeString(signPkHex)
		if err != nil {
			log.Fatal(err)
		}
		conf.SignPk = signPk
	}
	if encryptSkID := tomlConf.EncryptSkID; encryptSkID > 0 {
		conf.EncryptSkID = make([]byte, 8)
		binary.LittleEndian.PutUint64(conf.EncryptSkID, encryptSkID)
	} else if len(conf.EncryptSk) > 0 {
		hf, _ := blake2b.New(&blake2b.Config{
			Person: []byte(DomainStr),
			Size:   8,
		})
		hf.Write(conf.EncryptSk)
		encryptSkID := hf.Sum(nil)
		encryptSkID[7] &= 0x7f
		conf.EncryptSkID = encryptSkID
	}
	conf.TTL = DefaultTTL
	if ttl := tomlConf.TTL; ttl > 0 {
		conf.TTL = time.Duration(ttl) * time.Second
	}
	if signSkHex := tomlConf.SignSk; signSkHex != "" {
		signSk, err := hex.DecodeString(signSkHex)
		if err != nil {
			log.Fatal(err)
		}
		switch len(signSk) {
		case 32:
			if len(conf.SignPk) != 32 {
				log.Fatal("Public signing key required")
			}
			signSk = append(signSk, conf.SignPk...)
		case 64:
		default:
			log.Fatal("Unsupported length for the secret signing key")
		}
		conf.SignSk = signSk
	}
	conf.MaxClients = *maxClients
	conf.MaxLen = *maxLenMb * 1024 * 1024
	conf.Timeout = time.Duration(*timeout) * time.Second
	conf.DataTimeout = time.Duration(*dataTimeout) * time.Second
	conf.TrustedIPCount = uint64(float64(conf.MaxClients) * 0.1)
	if conf.TrustedIPCount < 1 {
		conf.TrustedIPCount = 1
	}
	confCheck(conf, *isServer)
	if *isServer {
		RunServer(conf)
	} else {
		RunClient(conf, *isCopy, *isMove)
	}
}
